### DISCUSSION TITLE
How to obfuscate entire project using pyarmor gen?

### DISCUSSION THREAD

**Discussion by willEternal8** [dashingsoft/pyarmor#1688] [Q&A]
Currently if i run `pyarmor gen -O obfuscated_project -r -i <project_name>` and then try to run main.py in `project_name` i get following error.
```
Traceback (most recent call last):
  File "/home/project_name/main.py", line 2, in <module>
    from .pyarmor_runtime_006205 import __pyarmor__
ImportError: attempted relative import with no known parent package
```

I have solved this by using python script to iterate through all the modules in project and run `pyarmor gen -O obfuscated_project -r -i <module_name>`. Is there a better way to do this?

Also my method fails if I have multiple sub modules.

**Comment by jondy** [comment]
Check the option `-i` and `--prefix` in the man page to understand how to use them.

**Reply by willEternal8** [reply]
Thanks @jondy, I went through the docs and tried few combinations of `--prefix` and `-i` but I'm still getting same error when i try to run the main.py which is located at `dist/project/main.py`.
```
    from .pyarmor_runtime_006205 import __pyarmor__
ImportError: attempted relative import with no known parent package
```

When obfuscated package by package (`pyarmor gen src/pkg1 src/pkg2 src/pkg3`) everything works but issue comes when I have multiple packages inside `pkg2`.

So is there a way to run `pyarmor` on `src` directly that wont give me `ImportError`


Thank you

**Comment by jondy** [comment]
The option `--prefix` is right for this case, and you must know what's relative import, please search `relative import` in Python documentation.

**Comment by rohitcoder** [comment]
@willEternal8 have you been able to solve this issue?

**Reply by willEternal8** [reply]
@rohitcoder not yet I'm still facing same error. But I was able to solve it by changing import statement from `from .pyarmor_runtime_006205 import __pyarmor__` to `from pyarmor_runtime_006205 import __pyarmor__`

**Reply by rohitcoder** [reply]
@jondy is this correct approach? I also noticed this was working. But i also noticed a problem where i did from `pkg.internals import system` (where pkg is my package name, which i get after pip3 install .) i was getting  some relative path import error.

**Reply by khadeom** [reply]
@rohitcoder The way I solved this issue was by obfuscating package by package and individual files when required using `pyarmor gen -r @filelist` as mentioned in documentation. https://pyarmor.readthedocs.io/en/stable/reference/man.html.

**Comment by jondy** [comment]
Suppose there are 3 package: src/pkg1, src/pkg2, src/pkg3
2 scripts: src/a.py, src/b.py
After obfuscated, runtime package is stored in the `pkg1`, just obfuscate them by this command:
```
pyarmor gen -r --prefix pkg1 src/a.py src/b.py src/pkg1 src/pkg2 src/pkg3
```

**Reply by neilyoung** [reply]
@jondy Is there something equivalent for older versions? 

```
pyarmor: error: argument : invalid choice: 'gen' (choose from 'obfuscate', 'o', 'licenses', 'l', 'pack', 'p', 'init', 'i', 'config', 'c', 'build', 'b', 'info', 'check', 'hdinfo', 'benchmark', 'capsule', 'register', 'download', 'runtime', 'help')
```

Using 6.8.1

**Comment by jondy** [comment]
Please check man page in Pyarmor 7 doc, I also need to check man page.
